### ISSUE TITLE
service-module: runtime management fails for systemd

### ISSUE THREAD

**Issue by herrsimon** [ansible/ansible#2003]
Hi,

the service module only checks for the 'service' binary or initscripts for service management (see lines 359-366 in module). On a pure systemd-installation, neither of these are present (instead one uses the 'systemctl' binary as well) and thus the service module fails. 

A quick and dirty solution would be to just insert more if statements but maybe one should separate the different service management systems into subclasses (inheriting either from the LinuxService subclass or directly from the Service class). 

I could do both and prefer the latter solution, but before investing time I would like to get the opinion of the maintainers here.  So what do you think?

**Comment by dhozac** [comment]
The problem is that it's quite common to have several systems at once, e.g. some native upstart/systemd services and some initscripts. service hides this fact. Another set of if statements looks right to me.

**Comment by mpdehaan** [comment]
+1

**Comment by mpdehaan** [comment]
@herrsimon is this something you would like to submit a pull request for?

**Comment by herrsimon** [comment]
Hello, 

another if-statement definitely doesn't seem right to me. Of course, this would fix the current problem with systemd but I bet you that the next problem of this kind  is just around the corner...

I came up with an idea for a  clean and easy to extend solution keeping the generic service module (i.e. away from the error-prone and inflexible ever growing if-statements to class inheritance and generic methods) but this basically means rewriting the whole module. Also, with keeping the module comes the limitation of not being able to support features specific to some init system (like for example runlevels).  Of course, one can again find a workaround, like the current 'arguments' keyword, but this is not the idea of a module. By the way, what should one do if there is more than one system that needs additional and different arguments? Ok, now you tell me about host variables and a $with_items loop but does this really simplify things?

A similar situation would arise with package management and I think there's a good reason why you chose to implement separate yum/apt/pacman modules instead of squeezing everything into one generic 'package' module.

I'm not sure if it wouldn't be nicer to have separate upstart/systemd/rcconf etc. modules, like it's the case with the package managers. One could then differentiate putting the hosts into groups. Another argument for this is that features not shared by all different init systems (runlevels, dependencies etc.) could be supported by the modules as well. 

So basically the question now is: A limited and possibly not feature complete generic service module or several modules, tailored to the different init systems. I would like to be convinced of one of these solutions before starting to code. So, what do you think?

Best wishes,

Simon

PS: Unrelated to this, why is the execute_command method not in module_common.py? I don't see why it's specific to the service module.

**Comment by mpdehaan** [comment]
Let's not discuss architecture.

All it needs is another if and we're good.

Next thing you know we'll have a
DynamicFacadeInjectorAbstractFactoryInstance and all goes to hell.

We don't want seperate upstart/etc as nobody remembers which service is
managed by what.

Mostly under RHT this just works with /sbin/service but there are some
exceptions.

**Comment by mpdehaan** [comment]
What OS are you using for a pure systemd installation BTW?

**Comment by herrsimon** [comment]
First of all: If my previous comment offended you in any way, I apologize. Avoiding complex if-statements is something I want to do instinctively, can't help it :-) So if you think the generic class is good as is, i'll just add the if-statements (which still leaves the enable/disable functionality non-idempotent). 

PS: The pure systemd installation is Archlinux.

**Comment by mpdehaan** [comment]
I'm planning on making various cleanups to the system module shortly which should be able to detect if a service is upstart, sysV, or systemd and should resolve this issue along with some others.

So I'd say hold off for now and this should be in place soon, and maybe you'll want to clean it up for arch once done

**Comment by mpdehaan** [comment]
ok, so this refactoring is now done.  If you are still having problems with pure-systemd, it will need some testing on those platforms, and we should be able to more easily include a fix.

If you can, please test and submit a minimal patch for what you find.

**Comment by mpdehaan** [comment]
@herrsimon 

Updated the latest devel for some better checks on when the service should be managed by a particular service management tool.

I did some tests on Fedora 18 and things seem pretty happy.  If you can test on your platform and let me know that would be great.
